### ISSUE TITLE
CVE-2021-45046 log4shell (cont'd)

### ISSUE THREAD

**Issue by sideeffect42** [eXist-db/exist#4125]
Since my original issue #4114 has been locked, I'm opening a new one asking for a hot fix also containing the patched patched version of log4j (2.16.0).

Would it be possible to remove the `Jndi*` classes from the bundled log4j JAR altogether as is recommended in the mitigation of [CVE-2021-45046](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-45046)?

**Comment by dizzzz** [comment]
We decided in the community discussion that for v5.3.1 the v2.15 version should be sufficient; actually the v2.16 version introduces such a significant functional change, that we think it can only applied in a major release of exist (v6.0). @adamretter  @duncdrum

**Comment by suchmaske** [comment]
Could you elaborate what parts would be affected? I have read the recommendation to swap the log4j version manually (also in older versions of exist). Will this cause trouble? @dizzzz

**Comment by duncdrum** [comment]
Since exist does not ship with the specific patterns required by the new CVE, `5.3.1` and soon `4.8.0` should not be affected by either log4shell `1.0` or `2.0`. However log4j2 `2.16` is already on develop. Users who wish to stay ahead are free to checkout the head of develop and to run / test their applications with `:latest`  docker images. 

Removing the JNDI classes is indeed a functional change requiring a new major release of exist `6.0.0`. It ll take some time for us to decide on a practical way about this, and to understand and test the full implications. With a number of `6.0.0` features and PRs in waiting for an extended period of time. It might make more sense to actually do a non hotfix release and target a swift release of `6.0.0` including more features. 

TL;DR: We are monitoring the situation. `5.3.1` should be save for production. The current head of develop offers users backwards incompatible changes but comes with log4j2 `2.16` for those who perfer to not rely on `2.15`

**Comment by sideeffect42** [comment]
Thanks for your explanations, @duncdrum.

> Removing the JNDI classes is indeed a functional change requiring a new major release of exist 6.0.0.

I understand, so I'll just do that locally as an additional hardening measure.

> With a number of 6.0.0 features and PRs in waiting for an extended period of time. It might make more sense to actually do a non hotfix release and target a swift release of 6.0.0 including more features.

I would definitely welcome a swifter release of 5.4.0/6.0.0, as there are some other things that have been in queue for some time (e.g. #4126/#3969).

**Comment by LaurentNoyon** [comment]
In the previous topic, duncdrum advices to replace %m with %m{nolookups} in log4j config file.
However, I saw that it might not be enough : https://stackoverflow.com/a/70391163/5831367

- Is the fix proposed by duncdrum still applicable ?
- Assuming it's not and I do not have the possibility to upgrade existdb, what would be the way to completely patch the vulnerability ? Completely disabling the logs would be fine for me.

**Comment by duncdrum** [comment]
For affected log4j versions < 2.15 `%m{nolookups}` is no longer deemed sufficient (still better than not doing it).  You would have to manually patch log4j, and if you application allows remove the JNDI class from log4j all together, by:
- make sure you have a current backup
- stopping exist 
- go to the the top level directory of your exist installation (aka `EXIST_HOME`) an open it in terminal
```shell
cd /your/path/to/exist
```
- Run: `zip -q -d lib/log4j-core-*.jar org/apache/logging/log4j/core/lookup/JndiLookup.class` 
- Restart exist

**Comment by sideeffect42** [comment]
Just a heads up, log4j 2.16.0 also isn't enough...
https://logging.apache.org/log4j/2.x/security.html#log4j-2.17.0

**Comment by adamretter** [comment]
This is fixed in eXist-db 5.4.0 and 6.0.0
